Issue #2228 "Grizzly NPE in DefaultFilterChain, this.serverRequest is…

carryel opened the pull request:
… null" (https://github.com/eclipse-ee4j/glassfish-grizzly/issues/2228)

Issue #2214 "NPE when closing OutputBuffer" (https://github.com/eclipse-ee4j/glassfish-grizzly/issues/2214)

+ Fixed an issue where a new request was unintentionally processed as a suspend request because it was referencing a recycled request.

+ Improvements to the use of recycled InputBuffer and OutputBuffer and concurrent access to them. When a request and response are completed and recycled, the associated ServerInputBuffer and ServerOutputBuffer are also recycled. However, if another thread references the InputBuffer and OutputBuffer, an NPE may occur when using these recycled instances. We consider the possibility of unintended usage after recycling, and the possibility of concurrent calls to recycle and use. I did not modify InputBuffer and OutputBuffer to be completely thread-safe; Improvements have been made only to the areas mentioned in the issues above.

+ Patched an NPE that occurred when calling toString() on a recycled HttpRequestPacket instance (preventing an NPE when obtaining the methodBytes of Method).

+ Patched for TransportFilter's exceptionOccurred() not propagating to filter chain (it seems that there was no apparent intention to prevent propagation).

avpinchuk commented on modules/http-server/src/main/java/org/glassfish/grizzly/http/server/io/ServerOutputBuffer.java:
```diff
@@ -17,23 +18,25 @@
     public void initialize(final Response response, final FilterChainContext ctx) {
-        super.initialize(response.getResponse(), response.isSendFileEnabled(), ctx);
         this.serverResponse = response;
+        super.initialize(response.getResponse(), response.isSendFileEnabled(), ctx);
```
Why volatile write happens before `super` initialization?

carryel replied:
@avpinchuk Thanks for reviewing this.
That's right. This order change doesn't currently have any significant implications for ServerOutputBuffer/OutputBuffer. It was simply done to maintain code consistency with ServerInputBuffer/InputBuffer.

https://github.com/carryel/grizzly/blob/439c8ec2eb92c54b1920c4156f44e3ee017aa829/modules/http-server/src/main/java/org/glassfish/grizzly/http/server/io/ServerInputBuffer.java#L41-L43

It is also to prevent mistakes in the future when ServerOutputBuffer's method is called again through super.initialize (like ServerInputBuffer/InputBuffer).

> super.initialize() -> ServerInputBuffer.someMethod()

https://github.com/carryel/grizzly/blob/439c8ec2eb92c54b1920c4156f44e3ee017aa829/modules/http/src/main/java/org/glassfish/grizzly/http/io/InputBuffer.java#L186-L189

https://github.com/carryel/grizzly/blob/439c8ec2eb92c54b1920c4156f44e3ee017aa829/modules/http-server/src/main/java/org/glassfish/grizzly/http/server/io/ServerInputBuffer.java#L81-L85

The reason for changing it to volatile is that, although rare, recycle(), which initializes the variable to null, can be called concurrently from a different thread than the one referencing the variable.

carryel commented:
This PR is a workaround patch rather than a fundamental fix, and I've confirmed that ignoring recycled requests and processing them as new requests resolves the issue properly. However, I haven't yet discovered the root cause of this issue and am not sure about it.
I think recycle for request and execution for new url should not happen at the same time within the same connection.

Therefore, I believe it would be best to further investigate the root cause before reviewing or applying this PR.
Let me look into it further.

carryel commented:
I think I found the root cause of the issue.
It appears to be an issue with the attribute holder that stores the request.

https://github.com/eclipse-ee4j/glassfish-grizzly/blob/227418b2fe2e3b1176d9bbe3ded9a42bfbea8d0e/modules/http-server/src/main/java/org/glassfish/grizzly/http/server/HttpServerFilter.java#L121

For example, after calling httpRequestInProgress.remove(context) and then calling httpRequestInProgress.get(context) to test for deletion, I observed in the log that in certain cases, the attribute value was not properly deleted and remained.

https://github.com/eclipse-ee4j/glassfish-grizzly/blob/227418b2fe2e3b1176d9bbe3ded9a42bfbea8d0e/modules/http-server/src/main/java/org/glassfish/grizzly/http/server/HttpServerFilter.java#L332

For the AttributeHolder issue, I'll continue the topic in a separate issue.

carryel commented:
This issue will no longer occur automatically once https://github.com/eclipse-ee4j/glassfish-grizzly/issues/2246 is resolved.
However, the main change is that ServerInputBuffer and ServerOutputBuffer have been improved to be more secure in multithreaded environments.
This is because these two InputBuffer and OutputBuffer can be accessed from multiple threads via NIOInputStream, NIOReader, NIOOutputStream, and NIOWriter.

carryel committed:
Issue #2228 "Grizzly NPE in DefaultFilterChain, this.serverRequest is null" (https://github.com/eclipse-ee4j/glassfish-grizzly/issues/2228)
Issue #2214 "NPE when closing OutputBuffer" (https://github.com/eclipse-ee4j/glassfish-grizzly/issues/2214)
+ Added FINE or FINEST log messages for HttpServerFilter's debugging.
+ If an exception occurs during the Suspended Request processing, it is handled gracefully on its own without being thrown.

+ Improvements to the use of recycled InputBuffer and OutputBuffer and concurrent access to them.
+ Handle exceptions more elegantly when NIOInputStream, NIOOuputStream, NIOReader, and NIOWriter are not initialized (or recycled).
When a request and response are completed and recycled, the associated ServerInputBuffer and ServerOutputBuffer are also recycled.
However, if another thread references the InputBuffer and OutputBuffer, an NPE may occur when using these recycled instances.
We consider the possibility of unintended usage after recycling, and the possibility of concurrent calls to recycle and use.
I did not modify InputBuffer and OutputBuffer to be completely thread-safe; Improvements have been made only to the areas mentioned in the issues above.

+ Added test cases that can cause NPEs that elegantly handle exceptions.

carryel commented:
Could you please review this PR when you have time?

While there appear to be a lot of changes, the logic itself remains largely unchanged, with only the following improvements:
- Checking for initialization: While initialization is always performed under normal circumstances, when async reads or writes occur in application code, Grizzly terminates the request and recycles it. However, an NPE can occur in the application that still holds the reference. While this is not a normal situation, we've improved the handling to ensure safer handling.
- Added FINE/FINEST log messages to facilitate future debugging.
- Replicated existing test cases to create an abnormal situation (normally, close() should be called before response.resume() in the application handler, but close() is intentionally called after resume() in testcases). I intentionally triggered an NPE and verified that this exception was handled gracefully.

PS) If possible, I'd like to see these improvements included in a new Grizzly release.
